SocketStats/main.cpp

```diff
@@ -0,0 +1,150 @@
+    }
+}
+
+
+// Network interface
+NetworkInterface *net;
+
+int main()
+{
+    Thread *thread = new Thread(osPriorityNormal1, 2048);
+    int remaining;
+    int rcount;
+    char *p;
+    char *buffer = new char[256];
+    nsapi_size_or_error_t result;
+
+    thread->start(print_stats);
+    // Bring up the ethernet interface
+    printf("Mbed OS Socket example\n");
+
+#ifdef MBED_MAJOR_VERSION
+    printf("Mbed OS version: %d.%d.%d\n\n", MBED_MAJOR_VERSION, MBED_MINOR_VERSION, MBED_PATCH_VERSION);
+#endif
+
+    net = NetworkInterface::get_default_instance();
+
+    if (!net) {
+        printf("Error! No network inteface found.\n");
+        return 0;
+    }
+
+    result = net->connect();
+    if (result != 0) {
+        printf("Error! net->connect() returned: %d\n", result);
+        return result;
+    }
+
+    // Show the network address
+    const char *ip = net->get_ip_address();
+    const char *netmask = net->get_netmask();
+    const char *gateway = net->get_gateway();
+    printf("IP address: %s\n", ip ? ip : "None");
+    printf("Netmask: %s\n", netmask ? netmask : "None");
+    printf("Gateway: %s\n", gateway ? gateway : "None");
+
+    // Open a socket on the network interface, and create a TCP connection to api.ipify.org
+    TCPSocket socket;
+    // Send a simple http request
+    char sbuffer[] = "GET / HTTP/1.1\r\nHost: api.ipify.org\r\nConnection: close\r\n\r\n";
+    nsapi_size_t size = strlen(sbuffer);
+
+    result = socket.open(net);
+    if (result != 0) {
+        printf("Error! socket.open() returned: %d\n", result);
+    }
+
+    result = socket.connect("api.ipify.org", 80);
+    if (result != 0) {
+        printf("Error! socket.connect() returned: %d\n", result);
+        goto DISCONNECT;
+    }
+
+    // Loop until whole request sent
+    while(size) {
+        result = socket.send(sbuffer+result, size);
+        if (result < 0) {
+            printf("Error! socket.send() returned: %d\n", result);
+            goto DISCONNECT;
+        }
+        size -= result;
+        printf("sent %d [%.*s]\n", result, strstr(sbuffer, "\r\n")-sbuffer, sbuffer);
+    }
+
+    // Receieve an HTTP response and print out the response line
+    remaining = 256;
+    rcount = 0;
+    p = buffer;
+    while (0 < (result = socket.recv(p, remaining))) {
+        p += result;
+        rcount += result;
+        remaining -= result;
+    }
+    if (result < 0) {
+        printf("Error! socket.recv() returned: %d\n", result);
+        goto DISCONNECT;
+    }
+    // the HTTP response code
+    printf("recv %d [%.*s]\n", rcount, strstr(buffer, "\r\n")-buffer, buffer);
+
+    // The api.ipify.org service also gives us the device's external IP address
+    p = strstr(buffer, "\r\n\r\n")+4;
+    printf("External IP address: %.*s\n", rcount-(p-buffer), p);
+    delete[] buffer;
+
+DISCONNECT:
+    // Close the socket to return its memory and bring down the network interface
+    socket.close();
+
+    ThisThread::sleep_for(SAMPLE_TIME);
```

Comment:
Minor thing: Is this sleep required here? Isn't it better to call thread->terminate() and thread->join() to sync with the thread exiting? It doesn't affect what the example is trying to demonstrate, so I would still approve this.

Reply:
Sleep is added to make sure we display the close event after socket close. We sleep in fetch thread https://github.com/ARMmbed/mbed-os-examples-docs_only/pull/12/files#diff-038ef0db0993b2992cbd5f531d8244a6R46 after each stats reporting and based on timing between socket.close() and terminate, there is little scope for thread to switch (based on where is in sleep) and print last stats showing socket was closed.

Adding sleep is just for getting print from reporting thread, socket close update will be done as soon as close is performed.